### PR TITLE
KAFKA-16536; Use BeginQuorumEpoch as leader heartbeat

### PR DESCRIPTION
**Pull request by ahuang98**
I've set the heartbeat cadence to half the fetch timeout for now (which would default to 1 second).
NonAcknowledgingVoters are no longer tracked explicitly - the heartbeat timer is initialized at 0ms so the leader will send begin quorum requests to every voter on startup, which will cover nonAcknowledgingVoters. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 1]
```diff
@@ -2132,15 +2132,30 @@ private long pollLeader(long currentTimeMs) {
...
-            currentTimeMs,
-            partitionState
-                .lastVoterSet()
-                .voterNodes(state.nonAcknowledgingVoters().stream(), channel.listenerName()),
-            this::buildBeginQuorumEpochRequest
-        );
+        long timeUntilNextBeginQuorumSend = state.timeUntilBeginQuorumEpochTimerExpires(currentTimeMs);
+
+        if (timeUntilNextBeginQuorumSend == 0) {
+            System.out.println("timeUntilNextBeginQuorumSend = 0");
```
Comment (dujian0068) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: `System.out.println` should be replaced by `logger`

Comment (ahuang98) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: thanks, left a debugging message in by accident. I'll change this to trace to reduce log spamming

Comment (ahuang98) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: I don't feel too strongly about having the trace messages I added - will let others leave their opinion

File: raft/src/main/java/org/apache/kafka/raft/internals/VoterSet.java [hunk 2]
```diff
@@ -56,6 +56,18 @@ final public class VoterSet {
...
+     * Returns the node information for all the voters
+     */
+    public Set<Node> voterNodes(ListenerName listenerName) {
+        Set<Node> allVoters = new HashSet<>();
+        for (VoterNode voter : voters.values()) {
+            voter.address(listenerName).ifPresent(address ->
+                allVoters.add(new Node(voter.voterKey().id(), address.getHostString(), address.getPort())));
+        }
+        return allVoters;
+    }
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/internals/VoterSet.java]: This should have the same semantic as `voterNodes(Stream, ListenerName)`. Throw an `IllegalArgumentException` is there is a voter that doesn't support the given listener.

Let's add a test for this new functionality.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 3]
```diff
@@ -2132,15 +2132,32 @@ private long pollLeader(long currentTimeMs) {
...
+            timeUntilNextBeginQuorumSend = maybeSendRequests(
+                currentTimeMs,
+                partitionState
+                    .lastVoterSet()
+                    .voterNodes(state.nonAcknowledgingVoters().stream(), channel.listenerName()),
+                this::buildBeginQuorumEpochRequest
+            );
+            logger.trace("Attempted to send BeginQuorumEpochRequest to non-acknowledging voters: {}. " +
+                "Request can be retried in {} ms", state.nonAcknowledgingVoters(), timeUntilNextBeginQuorumSend);
+        }
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: How about moving this algorithm to a private method? So that the changes to`pollLeader` become `maybeSendBeginQuorumEpochRequests(currentTimeMs);`.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 4]
```diff
@@ -2132,15 +2132,32 @@ private long pollLeader(long currentTimeMs) {
...
+                currentTimeMs,
+                partitionState
+                    .lastVoterSet()
+                    .voterNodes(channel.listenerName()),
+                this::buildBeginQuorumEpochRequest
+            );
+            state.resetBeginQuorumEpochTimer(currentTimeMs);
+            logger.trace("Attempted to send BeginQuorumEpochRequest as heartbeat to all voters. " +
+                "Request can be retried in {} ms", timeUntilNextBeginQuorumSend);
+        } else if (!state.nonAcknowledgingVoters().isEmpty()) {
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: I think that you can remove this `if` case if you set the `beginQuorumEpochTimer` timer to 0 (expired) when becoming leader for the first time.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 5]
```diff
@@ -495,6 +496,31 @@ public void testHandleBeginQuorumEpochAfterUserInitiatedResign() throws Exceptio
...
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters).build();
+
+        context.becomeLeader();
+        assertEquals(OptionalInt.of(localId), context.currentLeader());
+
+        // begin epoch requests should be sent out every beginQuorumEpochTimeoutMs
+        context.time.sleep(context.beginQuorumEpochTimeoutMs);
+        context.client.poll();
+        context.expectBeginEpoch(context.currentEpoch());
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: It looks `expectBeginEpoch` already `pollUntilRequests`. Having said that `expectBeginEpoch` is not precise enough. We need to check that the leader sent a BEGIN_QUORUM_EPOC request to all of the voters.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 6]
```diff
@@ -2132,15 +2132,32 @@ private long pollLeader(long currentTimeMs) {
...
-            this::buildBeginQuorumEpochRequest
-        );
+        long timeUntilNextBeginQuorumSend = state.timeUntilBeginQuorumEpochTimerExpires(currentTimeMs);
+
+        if (timeUntilNextBeginQuorumSend == 0) {
+            timeUntilNextBeginQuorumSend = maybeSendRequests(
+                currentTimeMs,
+                partitionState
+                    .lastVoterSet()
+                    .voterNodes(channel.listenerName()),
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: The leader should not send a BEGIN_QUORUM_EPOCH request to itself. Please write a detail test that checks this.

File: raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java [hunk 7]
```diff
@@ -789,6 +790,27 @@ public void testGrantVote(boolean isLogUpToDate) {
...
+        MockTime time = new MockTime();
+        int leader = localId;
+        int follower1 = 1;
+        long epochStartOffset = 10L;
+
+        Set<Integer> voterSet = mkSet(leader, follower1);
+        LeaderState<?> state = newLeaderState(voterSet, epochStartOffset);
+
+        long resetTime = time.milliseconds();
+        state.resetBeginQuorumEpochTimer(resetTime);
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java]: Before reseting the timer you should check that the expiration time is what you expect when becoming leader.

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java [hunk 8]
```diff
@@ -410,6 +411,7 @@ public void becomeLeader() throws Exception {
         time.sleep(electionTimeoutMs * 2L);
         expectAndGrantVotes(currentEpoch + 1);
         expectBeginEpoch(currentEpoch + 1);
+        client.poll();
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: Why this change?

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java [hunk 9]
```diff
@@ -444,13 +446,12 @@ private int localIdOrThrow() {
...
     }
 
-    private void expectBeginEpoch(int epoch) throws Exception {
+    public void expectBeginEpoch(int epoch) throws Exception {
         pollUntilRequest();
         for (RaftRequest.Outbound request : collectBeginEpochRequests(epoch)) {
             BeginQuorumEpochResponseData beginEpochResponse = beginEpochResponse(epoch, localIdOrThrow());
             deliverResponse(request.correlationId(), request.destination(), beginEpochResponse);
         }
-        client.poll();
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: Why this change?

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 10]
```diff
@@ -495,6 +496,31 @@ public void testHandleBeginQuorumEpochAfterUserInitiatedResign() throws Exceptio
...
+
+        context.client.poll();
+        assertThrows(AssertionFailedError.class, () -> context.expectBeginEpoch(context.currentEpoch()));
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: What is this testing?

**Review by jsancio — COMMENTED** [review]
Thanks for the changes @ahuang98

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java (hunk 10 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: I'll fix this - the point is to check we're not sending requests before the timeout is expired

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java (hunk 8 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: I was thinking `expectBeginEpoch` should only check for beginEpoch requests and we should leave it to the test to call poll when needed

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java (hunk 4 above)
Comment (ahuang98) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: got it, if the non-ack voters miss the beginQuorumEpoch request the first time - we'll rely on the heartbeat to retry requests?

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java (hunk 8 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: I won't need `expectBeginEpoch` in my tests anymore, so I'll revert the change.

**Comment by ahuang98** [comment]
@jsancio Thanks for the detailed review. Some comments seemed less applicable after addressing other comments, feel free to point out if I shouldn't have reverted some changes.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 11]
```diff
@@ -2096,6 +2096,26 @@ private long maybeAppendBatches(
...
+        LeaderState<T> state,
+        long currentTimeMs
+    ) {
+        long timeUntilNextBeginQuorumSend = state.timeUntilBeginQuorumEpochTimerExpires(currentTimeMs);
+        if (timeUntilNextBeginQuorumSend == 0) {
+            timeUntilNextBeginQuorumSend = maybeSendRequests(
+                currentTimeMs,
+                partitionState
+                    .lastVoterSet()
+                    .voterNodes(state.votersExcludingLeader().stream(), channel.listenerName()),
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: The latest voters are in `partitionState.lastVoterSet()`. It is odd that `KafkaRaftClient` needs to query `LeaderState` to exclude itself from the `VoterSet`. How about `voters.voterKeys().stream().filter(!self)`?

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java [hunk 12]
```diff
@@ -2096,6 +2096,26 @@ private long maybeAppendBatches(
...
+                this::buildBeginQuorumEpochRequest
+            );
+            state.resetBeginQuorumEpochTimer(currentTimeMs);
+            logger.trace("Attempted to send BeginQuorumEpochRequest as heartbeat to all voters. " +
+                "Request can be retried in {} ms", timeUntilNextBeginQuorumSend);
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: This doesn't match the formatting used in this file or module:
```java
            logger.trace(
                "Attempted to send BeginQuorumEpochRequest as heartbeat to all voters. " +
                "Request can be retried in {} ms",
                timeUntilNextBeginQuorumSend
            );
```

File: raft/src/main/java/org/apache/kafka/raft/LeaderState.java [hunk 13]
```diff
@@ -220,7 +234,14 @@ public Uuid localDirectoryId() {
         return localDirectoryId;
     }
 
-    public Set<Integer> nonAcknowledgingVoters() {
+    public Set<Integer> votersExcludingLeader() {
+        Set<Integer> voters = new HashSet<>(voterStates.keySet());
+        voters.remove(localId);
+        return voters;
+    }
```
Comment (jsancio) [inline raft/src/main/java/org/apache/kafka/raft/LeaderState.java]: See my other comment but you should be able to remove this method.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 14]
```diff
@@ -497,6 +497,32 @@ public void testHandleBeginQuorumEpochAfterUserInitiatedResign() throws Exceptio
...
+
+        RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters).build();
+
+        context.becomeLeader();
+        assertEquals(OptionalInt.of(localId), context.currentLeader());
+
+        // begin epoch requests should be sent out every beginQuorumEpochTimeoutMs
+        context.time.sleep(context.beginQuorumEpochTimeoutMs);
+        context.client.poll();
+        context.assertSentBeginQuorumEpochRequest(context.currentEpoch(), 2);
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: Does this method check that it was sent to `remoteId1` and `remoteId2`? If not, we should check that begin quorum epoch request was sent to all of the remote voters.

File: raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java [hunk 15]
```diff
@@ -866,6 +872,27 @@ public void testGrantVote(boolean isLogUpToDate) {
...
+    public void testBeginQuorumEpochTimer() {
+        int leader = localId;
+        int follower1 = 1;
+        long epochStartOffset = 10L;
+
+        Set<Integer> voterSet = mkSet(leader, follower1);
+        LeaderState<?> state = newLeaderState(voterSet, epochStartOffset);
+        assertEquals(0, state.timeUntilBeginQuorumEpochTimerExpires(time.milliseconds()));
+
+        long resetTime = time.milliseconds();
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java]: `time.milliseconds` is mocked so it will return the same value as long as you don't call `time.sleep(long)`.

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java [hunk 16]
```diff
@@ -628,7 +629,7 @@ void deliverResponse(int correlationId, Node source, ApiMessage response) {
     RaftRequest.Outbound assertSentBeginQuorumEpochRequest(int epoch, int numBeginEpochRequests) {
         List<RaftRequest.Outbound> requests = collectBeginEpochRequests(epoch);
         assertEquals(numBeginEpochRequests, requests.size());
-        return requests.get(0);
+        return !requests.isEmpty() ? requests.get(0) : null;
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: Let's avoid returning `null` from "public" methods.

We normally don't test the absence of a side effect because those checks are technically infinite. Maybe we can expose `collectBeginEpochRequests` as package private and check the size of the list.

**Review by jsancio — COMMENTED** [review]
Thanks for the changes @ahuang98

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java (hunk 12 above)
Comment (ahuang98) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: hm, looks like the formatting is all over the place for log statements in this file - I'll change this one though to match the proposed format

File: raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java (hunk 15 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java]: I can add a sleep before this, the main point was just to test the reset functionality though. 
Or are you thinking it's unnecessary to define this variable? I was thinking it makes it a bit more obvious to the reader that when you reset at a specific time X, the time left at time X is the total timeout.

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java (hunk 16 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: only one test actually needs the request data - I'll keep this helper to minimize redundancy of assertions, and rewrite the one test to grab the request data from collectBeginEpochRequests directly.

File: raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java (hunk 15 above)
Comment (jsancio) [reply raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java]: It is unnecessary to save the value in a variable.

File: raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java (hunk 12 above)
Comment (jsancio) [reply raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java]: Yes. Going forward, we want to use the formatting shown above. There may be lines that don't follow this formatting but I try to fix them as I see them.

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java [hunk 17]
```diff
@@ -625,10 +626,10 @@ void deliverResponse(int correlationId, Node source, ApiMessage response) {
         channel.mockReceive(new RaftResponse.Inbound(correlationId, response, source));
     }
 
-    RaftRequest.Outbound assertSentBeginQuorumEpochRequest(int epoch, int numBeginEpochRequests) {
+    void assertSentBeginQuorumEpochRequest(int epoch, int numBeginEpochRequests, Set<Integer> destinationIds) {
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: `numBeginEpochRequests` is implicit from the size of the `destinationIds` set. In other words, `assertEquals(numBeginEpochRequests, destinationIds)` is always true so you can remove the parameter and replace `numBeginEpochRequests` in this method with `destinationIds.size()`.

**Review by jsancio — COMMENTED** [review]
The changes look good to me. Just one suggestion.

This build configuration in Jenkins never finished: JDK 21 and Scala 2.13

File: raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java (hunk 15 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/LeaderStateTest.java]: I'll remove it - just thought it would make it extremely clear the times are supposed to be equal for that check

File: raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java (hunk 17 above)
Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/RaftClientTestContext.java]: ~I was thinking this would also check that we don't send more than one request to a destination~ Oops other way around, allow us to check if we send multiple to a destination but that definitely isn't needed

**Comment by ahuang98** [comment]
@jsancio there were quite a few conflicts from the merge and I think one of your comments needs revisiting - specifically  
> The latest voters are in partitionState.lastVoterSet(). It is odd that KafkaRaftClient needs to query LeaderState to exclude itself from the VoterSet. How about voters.voterKeys().stream().filter(!self)?

Since `localId()` was removed from a prior change, do you think it makes sense to add a public `localReplicaKey()` to LeaderState just so `maybeSendBeginQuorumEpochRequest` can filter on it, or should I add back a helper into LeaderState along the lines of `public Set<Integer> votersExcludingLeader()`?

**Comment by jsancio** [comment]
> @jsancio there were quite a few conflicts from the merge and I think one of your comments needs revisiting - specifically
> 
> > The latest voters are in partitionState.lastVoterSet(). It is odd that KafkaRaftClient needs to query LeaderState to exclude itself from the VoterSet. How about voters.voterKeys().stream().filter(!self)?
> 
> Since `localId()` was removed from a prior change, do you think it makes sense to add a public `localReplicaKey()` to LeaderState just so `maybeSendBeginQuorumEpochRequest` can filter on it, or should I add back a helper into LeaderState along the lines of `public Set<Integer> votersExcludingLeader()`?

You can use `QuorumState.localIdOrThrow()` since you know that you are the leader and voters/leaders must have a replica id.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java [hunk 18]
```diff
@@ -546,6 +546,32 @@ public void testHandleBeginQuorumEpochAfterUserInitiatedResign(
...
+        // begin epoch requests should be sent out every beginQuorumEpochTimeoutMs
+        context.time.sleep(context.beginQuorumEpochTimeoutMs);
+        context.client.poll();
+        context.assertSentBeginQuorumEpochRequest(context.currentEpoch(), Utils.mkSet(remoteId1, remoteId2));
+
+        context.time.sleep(context.beginQuorumEpochTimeoutMs / 2);
+        context.client.poll();
+        context.assertSentBeginQuorumEpochRequest(context.currentEpoch(), Utils.mkSet());
+
+        context.time.sleep(context.beginQuorumEpochTimeoutMs / 2);
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: `2 * (context.beginQuorumEpochTimeoutMs / 2)` when `context.beginQuorumEpochTimeoutMs` is an `int` is not always greater than or equal to `context.beginQuorumEpochTimeoutMs`. For example, `99 / 2 = 49` and `98` (49 * 2) is not greater than or equal to `99`.

You technically want:
```java
        int partialDelay = context.beginQuorumEpochTimeoutMs / 2;
        context.time.sleep(partialDelay);
        context.client.poll();
        context.assertSentBeginQuorumEpochRequest(context.currentEpoch(), Utils.mkSet());

        context.time.sleep(context.beginQuorumEpochTimeoutMs - partialDelay);
```

Comment (ahuang98) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientTest.java]: sounds good

**Comment by ahuang98** [comment]
@jsancio I noticed you changed the title to use a semicolon instead - I always thought that was a typo when I saw commits with this on trunk. Is that the convention?

**Comment by jsancio** [comment]
> @jsancio I noticed you changed the title to use a semicolon instead - I always thought that was a typo when I saw commits with this on trunk. Is that the convention?

I guess there is no standard. If you look at the earlier PRs they used `;` but I guess `:` is also correct. Feel free to change it back.

**Review by jsancio — APPROVED** [review]
LGTM. Thanks for the improvement @ahuang98
